Rules/ReviewUnusedParameter.cs

```diff
@@ -51,6 +51,29 @@ public IEnumerable<DiagnosticRecord> AnalyzeScript(Ast ast, string fileName)
                     continue;
                 }
 
+                // bail out if $MyInvocation.BoundParameters used
+                if (variableCount.ContainsKey("MyInvocation"))
+                {
+                    bool isBound = false;
+                    IEnumerable<Ast> memExpAst = scriptBlockAst.FindAll(oneAst => oneAst is MemberExpressionAst, false);
+                    foreach (MemberExpressionAst memberExpressionAst in memExpAst)
+                    {                        
+                        if (memberExpressionAst.Expression is VariableExpressionAst &&
+                            ((VariableExpressionAst)memberExpressionAst.Expression).VariablePath.UserPath == "MyInvocation" &&
+                            memberExpressionAst.Member is StringConstantExpressionAst &&
+                            ((StringConstantExpressionAst)memberExpressionAst.Member).Value == "BoundParameters")
+                        {
+                            isBound = true;
+                            break;
+                        }                        
+                    }
+
+                    if (isBound)
+                    {
+                        continue;
+                    }
+                }

```

Comment:
I know this PR has been closed, but rather than having a simple condition in the `FindAll` test and then processing the results, it's better to have a constraining condition in the `FindAll`, since `FindAll` actually builds a list of all the things that satisfy its predicate, meaning that having a stronger predicate will lead to less allocation.

Also, creating a static method and using that for the predicate will prevent the need for allocating a closure for the predicate.

For example:

```csharp
private static bool IsBoundParametersReference(Ast ast)
{
    return ast is MemberExpressionAst memberAst
        && memberAst.Expression is VariableExpressionAst variableAst
        && variableAst.VariablePath.UserPath.Equals("MyInvocation", StringComparison.OrdinalIgnoreCase)
        && memberAst.Member is StringConstantExpressionAst memberStringAst
        && memberStringAst.Equals("BoundParameters", StringComparison.OrdinalIgnoreCase);
}

...

// If $MyInvocation.BoundParameters is used, we must assume the parameter has been used
if (scriptBlockAst.FindAll(IsBoundParametersReference, false).Any())
{
    continue;
}
```

Even nicer, you can merge the logic for detecting `$PSBoundParameters` into the same method too if you like

Reply:
nice